File: src/group.rs
//! Группа воркеров с автоматической ротацией кредов.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use rand::Rng;
use tokio::sync::mpsc;

use crate::config::TurnParams;
use crate::creds;
use crate::dispatcher::Dispatcher;
use crate::session;
use crate::stats::Stats;

pub async fn worker_group(
    group_id: usize,
    hash_index: usize,
    tp: TurnParams,
    peer: std::net::SocketAddr,
    dispatcher: Arc<Dispatcher>,
    local_port: String,
    use_tcp: bool,
    get_config: bool,
    config_tx: Option<mpsc::Sender<String>>,
    worker_ids: Vec<usize>,
    initial_delay: Duration,
    cycle_duration: Duration,
    global_shutdown: Arc<AtomicBool>,
    global_pause: Arc<AtomicBool>,
    stats: Arc<Stats>,
    app_id: String,
    app_secret: String,
    device_id: String,
    password: String,
) {
    // Фазовый сдвиг
    if !initial_delay.is_zero() {
        log::info!(
            "[ГРУППА #{}] Фазовый сдвиг: {:?}",
            group_id,
            initial_delay
        );
        tokio::select! {
            _ = tokio::time::sleep(initial_delay) => {}
            _ = wait_shutdown(&global_shutdown) => { return; }
        }
    }

    let mut cycle_number = 0u64;
    let mut config_sent = !get_config;
    let mut prev_batch: Option<(Arc<AtomicBool>, Vec<tokio::task::JoinHandle<()>>)> = None;

    loop {
        if global_shutdown.load(Ordering::Relaxed) {
            // Убиваем предыдущий батч если есть
            if let Some((bs, handles)) = prev_batch.take() {
                bs.store(true, Ordering::Relaxed);
                for h in handles { let _ = tokio::time::timeout(Duration::from_secs(3), h).await; }
            }
            return;
        }

        // Doze-mode пауза: убиваем воркеров и ждём RESUME
        if global_pause.load(Ordering::Relaxed) {
            if let Some((bs, handles)) = prev_batch.take() {
                bs.store(true, Ordering::Relaxed);
                for h in handles { let _ = tokio::time::timeout(Duration::from_secs(3), h).await; }
            }
            log::info!("[ГРУППА #{}] Пауза (Doze)", group_id);
            loop {
                if global_shutdown.load(Ordering::Relaxed) { return; }
                if !global_pause.load(Ordering::Relaxed) {
                    log::info!("[ГРУППА #{}] Возобновление — новые креды", group_id);
                    break;
                }
                tokio::time::sleep(Duration::from_secs(1)).await;
            }
        }

        // Получаем креды ДО убийства старого батча (бесшовная ротация)
        let hash = tp.hashes[hash_index % tp.hashes.len()].clone();

        log::info!(
            "[ГРУППА #{}] Цикл {}: запрос кредов (хеш: {}...)",
            group_id,
            cycle_number,
            &hash[..std::cmp::min(8, hash.len())]
        );

        let creds_result = creds::get_creds_with_fallback(
            &tp,
            &hash,
            &global_shutdown,
            &stats,
            &app_id,
            &app_secret,
        )
        .await;

        let creds = match creds_result {
            Ok(c) => c,
            Err(e) => {
                if global_shutdown.load(Ordering::Relaxed) {
                    return;
                }
                log::error!("[ГРУППА #{}] Ошибка кредов: {}", group_id, e);
                tokio::select! {
                    _ = tokio::time::sleep(Duration::from_secs(30)) => {}
                    _ = wait_shutdown(&global_shutdown) => { return; }
                }
                continue;
            }
        };

        log::info!(
            "[ГРУППА #{}] Креды OK, TURN: {}, {} воркеров",
            group_id,
            creds.turn_url,
            worker_ids.len()
        );

        // ТЕПЕРЬ убиваем старый батч (креды уже готовы — минимальный простой)
        if let Some((bs, handles)) = prev_batch.take() {
            bs.store(true, Ordering::Relaxed);
            for h in handles {
                let _ = tokio::time::timeout(Duration::from_secs(3), h).await;
            }
        }

        // Создаём новый batch
        let batch_shutdown = Arc::new(AtomicBool::new(false));
        let config_needed = Arc::new(AtomicBool::new(!config_sent));

        let mut handles = Vec::new();

        for (i, &wid) in worker_ids.iter().enumerate() {
            let tp_c = tp.clone();
            let creds_c = creds.clone();
            let disp_c = dispatcher.clone();
            let lp = local_port.clone();
            let batch_sd = batch_shutdown.clone();
            let global_sd = global_shutdown.clone();
            let st = stats.clone();
            let dev = device_id.clone();
            let pwd = password.clone();
            let cn = config_needed.clone();

            let ctx = if !config_sent {
                config_tx.clone()
            } else {
                None
            };

            // Stagger: 500мс между воркерами
            let worker_delay = Duration::from_millis((i as u64) * 500);

            let handle = tokio::spawn(async move {
                if !worker_delay.is_zero() {
                    tokio::select! {
                        _ = tokio::time::sleep(worker_delay) => {}
                        _ = wait_shutdown_combined(&batch_sd, &global_sd) => { return; }
                    }
                }

                let should_get_config = cn
                    .compare_exchange(true, false, Ordering::SeqCst, Ordering::Relaxed)
                    .is_ok();

                // Retry loop: воркер переподключается при ошибке вместо тихой смерти
                let mut attempt = 0u32;
                loop {
                    if batch_sd.load(Ordering::Relaxed) || global_sd.load(Ordering::Relaxed) {
                        return;
                    }

                    let combined_shutdown = Arc::new(AtomicBool::new(false));

                    let cs = combined_shutdown.clone();
                    let bs = batch_sd.clone();
                    let gs = global_sd.clone();
                    let shutdown_watcher = tokio::spawn(async move {
                        loop {
                            if bs.load(Ordering::Relaxed) || gs.load(Ordering::Relaxed) {
                                cs.store(true, Ordering::Relaxed);
                                return;
                            }
                            tokio::time::sleep(Duration::from_millis(100)).await;
                        }
                    });

                    // Конфиг запрашиваем только при первой попытке
                    let get_conf = should_get_config && attempt == 0;

                    let result = session::run_session(
                        &tp_c,
                        peer,
                        &disp_c,
                        &lp,
                        use_tcp,
                        get_conf,
                        &ctx,
                        wid,
                        &creds_c,
                        &dev,
                        &pwd,
                        &combined_shutdown,
                        &st,
                    )
                    .await;

                    shutdown_watcher.abort();

                    match result {
                        Ok(()) => {
                            // Штатное завершение (таймаут чтения) — ретрай
                        }
                        Err(e) => {
                            if batch_sd.load(Ordering::Relaxed) || global_sd.load(Ordering::Relaxed) {
                                return;
                            }
                            // Ошибка квоты или кредов — не ретраим
                            if e.contains("TURN квота") || e.contains("Креды мертвы") || e.contains("FATAL_AUTH") {
                                log::error!("[ВОРКЕР #{}] Фатальная ошибка: {}", wid, e);
                                return;
                            }
                            attempt += 1;
                            log::warn!("[ВОРКЕР #{}] Ошибка (попытка {}): {}", wid, attempt, e);
                        }
                    }

                    // Пауза перед ретраем с джиттером (от 5 до 15 сек), чтобы размазать переподключения
                    if batch_sd.load(Ordering::Relaxed) || global_sd.load(Ordering::Relaxed) {
                        return;
                    }
                    let r_delay = rand::Rng::gen_range(&mut rand::thread_rng(), 5..16);
                    let retry_delay = Duration::from_secs(r_delay);
                    tokio::select! {
                        _ = tokio::time::sleep(retry_delay) => {}
                        _ = wait_shutdown_combined(&batch_sd, &global_sd) => { return; }
                    }
                }
            });

            handles.push(handle);
        }

        if !config_sent && config_needed.load(Ordering::Relaxed) == false {
            config_sent = true;
        }

        // Сохраняем батч для бесшовной ротации
        prev_batch = Some((batch_shutdown.clone(), handles));

        // Ждём TTL
        tokio::select! {
            _ = tokio::time::sleep(cycle_duration) => {
                log::info!("[ГРУППА #{}] TTL {:?} истёк, ротация", group_id, cycle_duration);
            }
            _ = wait_shutdown(&global_shutdown) => {
                if let Some((bs, handles)) = prev_batch.take() {
                    bs.store(true, Ordering::Relaxed);
                    for h in handles {
                        let _ = tokio::time::timeout(Duration::from_secs(5), h).await;
                    }
                }
                return;
            }
        }

        cycle_number += 1;

        if !config_sent {
            if config_needed.load(Ordering::Relaxed) == false {
                config_sent = true;
            }
        }
    }
}

async fn wait_shutdown(shutdown: &Arc<AtomicBool>) {
    loop {
        if shutdown.load(Ordering::Relaxed) {
            return;
        }
        tokio::time::sleep(Duration::from_millis(200)).await;
    }
}

async fn wait_shutdown_combined(a: &Arc<AtomicBool>, b: &Arc<AtomicBool>) {
    loop {
        if a.load(Ordering::Relaxed) || b.load(Ordering::Relaxed) {
            return;
        }
        tokio::time::sleep(Duration::from_millis(200)).await;
    }
}

File: go_client/group.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"net"
	"strings"
	"sync/atomic"
	"time"
)

const (
	workersPerGroup = 12
	cycleSecs       = 840 // 
	phaseStepSec    = 16  // 
)

// WorkerGroup:
// бесшовная ротация: получить новые креды → запустить новый батч → убить старый.
func WorkerGroup(
	ctx context.Context,
	groupID int,
	hashIndex int,
	tp *TurnParams,
	peer *net.UDPAddr,
	d *Dispatcher,
	localPort string,
	useUDP bool,
	getConfig bool,
	configCh chan<- string,
	workerIDs []int,
	initialDelay time.Duration,
	cycleDuration time.Duration,
	pauseFlag *int32,
	deviceID, password string,
	stats *Stats,
) {
	// Фазовый сдвиг
	if initialDelay > 0 {
		log.Printf("[ГРУППА #%d] Фазовый сдвиг: %v", groupID, initialDelay)
		select {
		case <-time.After(initialDelay):
		case <-ctx.Done():
			return
		}
	}

	cycleNumber := 0
	configSent := !getConfig

	// Предыдущий батч
	var prevCancel context.CancelFunc
	var prevDoneChs []chan struct{}

	killBatch := func() {
		if prevCancel != nil {
			prevCancel()
			for _, ch := range prevDoneChs {
				select {
				case <-ch:
				case <-time.After(3 * time.Second):
				}
			}
			prevCancel = nil
			prevDoneChs = nil
		}
	}
	defer killBatch()

	for {
		if ctx.Err() != nil {
			return
		}

		// Doze-mode пауза: убиваем воркеров и ждём RESUME
		if atomic.LoadInt32(pauseFlag) != 0 {
			killBatch()
			log.Printf("[ГРУППА #%d] Пауза (Doze)", groupID)
			for {
				if ctx.Err() != nil {
					return
				}
				if atomic.LoadInt32(pauseFlag) == 0 {
					log.Printf("[ГРУППА #%d] Возобновление — новые креды", groupID)
					break
				}
				time.Sleep(1 * time.Second)
			}
		}

		// Получаем креды ДО убийства старого батча (бесшовная ротация)
		hash := tp.Hashes[hashIndex%len(tp.Hashes)]
		shortHash := hash
		if len(shortHash) > 8 {
			shortHash = shortHash[:8]
		}
		log.Printf("[ГРУППА #%d] Цикл %d: запрос кредов (хеш: %s...)", groupID, cycleNumber, shortHash)

		creds, err := GetCredsWithFallback(ctx, tp, hash, stats)
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			log.Printf("[ГРУППА #%d] Ошибка кредов: %v", groupID, err)
			select {
			case <-time.After(30 * time.Second):
			case <-ctx.Done():
				return
			}
			continue
		}

		log.Printf("[ГРУППА #%d] Креды OK, TURN: %v, %d воркеров", groupID, creds.TurnURLs, len(workerIDs))

		// ТЕПЕРЬ убиваем старый батч (креды уже готовы — минимальный простой)
		killBatch()

		// Создаём новый batch
		batchCtx, batchCancel := context.WithCancel(ctx)
		var configNeeded int32
		if !configSent {
			configNeeded = 1
		}

		doneChs := make([]chan struct{}, len(workerIDs))

		for i, wid := range workerIDs {
			doneCh := make(chan struct{})
			doneChs[i] = doneCh

			// Stagger: 500мс между воркерами
			workerDelay := time.Duration(i) * 500 * time.Millisecond

			go func(wid int, delay time.Duration, doneCh chan struct{}) {
				defer close(doneCh)

				if delay > 0 {
					select {
					case <-time.After(delay):
					case <-batchCtx.Done():
						return
					}
				}

				shouldGetConfig := atomic.CompareAndSwapInt32(&configNeeded, 1, 0)

				// Retry loop: воркер переподключается при ошибке
				attempt := 0
				for {
					if batchCtx.Err() != nil {
						return
					}

					getConf := shouldGetConfig && attempt == 0
					var cc chan<- string
					if getConf && !configSent {
						cc = configCh
					}

					sessErr := RunSession(batchCtx, tp, peer, d, localPort, useUDP,
						getConf, cc, wid, creds, deviceID, password, stats)

					if sessErr != nil {
						if batchCtx.Err() != nil {
							return
						}
						errStr := sessErr.Error()

						// Дописываем понятные пояснения для типичных ошибок со стороны балансировщиков ВК
						errStrLower := strings.ToLower(errStr)
						if strings.Contains(errStrLower, "attribute not found") ||
							strings.Contains(errStrLower, "rate limit") ||
							strings.Contains(errStrLower, "flood control") ||
							strings.Contains(errStrLower, "ip mismatch") ||
							strings.Contains(errStrLower, "error 29") {
							errStr += " (ошибка со стороны ВК)"
						}

						// Фатальные ошибки — не ретраим
						if strings.Contains(errStr, "TURN квота") ||
							strings.Contains(errStr, "хеш мёртв") ||
							strings.Contains(errStr, "FATAL_AUTH") {
							log.Printf("[ВОРКЕР #%d] Фатальная ошибка: %s", wid, errStr)
							return
						}
						attempt++
						log.Printf("[ВОРКЕР #%d] Ошибка (попытка %d): %s", wid, attempt, errStr)
					}

					if batchCtx.Err() != nil {
						return
					}

					// Пауза перед ретраем с джиттером 5-15 сек
					retryDelay := time.Duration(5+rand.Intn(11)) * time.Second
					select {
					case <-time.After(retryDelay):
					case <-batchCtx.Done():
						return
					}
				}
			}(wid, workerDelay, doneCh)
		}

		if !configSent && atomic.LoadInt32(&configNeeded) == 0 {
			configSent = true
		}

		// Сохраняем батч для бесшовной ротации
		prevCancel = batchCancel
		prevDoneChs = doneChs

		// Ждём TTL
		select {
		case <-time.After(cycleDuration):
			log.Printf("[ГРУППА #%d] TTL %v истёк, ротация", groupID, cycleDuration)
		case <-ctx.Done():
			return
		}

		cycleNumber++
		if !configSent && atomic.LoadInt32(&configNeeded) == 0 {
			configSent = true
		}
	}
}

// ComputeGroupTiming вычисляет фазу запуска
func ComputeGroupTiming(groupIndex int) (initialDelay, cycle time.Duration) {
	phase := time.Duration(groupIndex*phaseStepSec) * time.Second
	return phase, time.Duration(cycleSecs) * time.Second
}

// ValidateSchedule проверяет расписание
func ValidateSchedule(numGroups int) {
	maxPhase := (numGroups - 1) * phaseStepSec
	if maxPhase >= cycleSecs {
		log.Printf("⚠ Фазовый сдвиг последней группы (%dс) >= цикла (%dс)", maxPhase, cycleSecs)
	}
	log.Printf("Расписание: %d групп, цикл=%dс (%dмин), сдвиг=%dс",
		numGroups, cycleSecs, cycleSecs/60, phaseStepSec)
}

func phaseDuration(secs int) time.Duration {
	return time.Duration(secs) * time.Second
}

// ParseHashes — парсит строку хешей
func ParseHashes(raw string) []string {
	var result []string
	for _, h := range strings.Split(raw, ",") {
		h = strings.TrimSpace(h)
		if idx := strings.IndexAny(h, "/?#"); idx != -1 {
			h = h[:idx]
		}
		if h != "" {
			result = append(result, h)
		}
	}
	return result
}

// TurnParams — конфигурация TURN
type TurnParams struct {
	Host          string
	Port          string
	Hashes        []string
	SecondaryHash string
	Sni           string
}

// Unused import suppressor
var _ = fmt.Sprintf
